File: Top200/24-SwapPairs.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummy := &ListNode{Val: -1}
	prev := dummy
	prev.Next = head
	curr := head

	for curr != nil && curr.Next != nil {
		prev.Next = curr.Next
		curr.Next = prev.Next.Next
		prev.Next.Next = curr
		prev = curr
		curr = curr.Next
	}

	return dummy.Next
}

func swapPairs2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	ret := swapPairs2(head.Next.Next)
	temp := head.Next

	head.Next.Next = head
	head.Next = ret

	return temp
}

File: Top200/24-SwapPairs.java
package ACFirst;

class SwapPairs {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        ListNode curr = head;
        prev.next = head;

        while (curr != null && curr.next != null) {
            prev.next = curr.next;
            curr.next = prev.next.next;
            prev.next.next = curr;
            prev = curr;
            curr = curr.next;
        }

        return dummy.next;
    }

    public ListNode swapPairs2(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode ret = this.swapPairs2(head.next.next);
        ListNode temp = head.next;

        head.next.next = head;
        head.next = ret;

        return temp;
    }
}
